base/math.jl

```diff
@@ -536,38 +536,56 @@ Stacktrace:
 [...]
 ```
 """
-hypot(x::Number, y::Number) = hypot(promote(x, y)...)
-function hypot(x::T, y::T) where T<:Number
-    ax = abs(x)
-    ay = abs(y)
-    if ax < ay
-        ax, ay = ay, ax
-    end
+function hypot(x::Number, y::Number)
+	#= Keep specialized two argument version, because it is faster
+	than generic varargs version below =#
+	ay, ax = minmax(abs(x), abs(y))
     if iszero(ax)
         r = ay / oneunit(ax)
     else
         r = ay / ax
     end
-
     rr = ax * sqrt(1 + r * r)
 
-    # Use type of rr to make sure that return type is the same for
-    # all branches
-    if isnan(r)
-        isinf(ax) && return oftype(rr, Inf)
-        isinf(ay) && return oftype(rr, Inf)
-        return oftype(rr, r)
-    else
-        return rr
-    end
+	if isnan(r)
+		if isinf(x) || isinf(y)
+			return oftype(rr, Inf)
+		else
+			return oftype(rr, NaN)
+		end
+   	else
+   		return rr
+   	end
 end
-
 """
     hypot(x...)
 
 Compute the hypotenuse ``\\sqrt{\\sum x_i^2}`` avoiding overflow and underflow.
 """
-hypot(x::Number...) = sqrt(sum(abs2(y) for y in x))
+hypot(x::Number...) = _hypot(promote(x...))
+function _hypot(x::NTuple{N,<:Number}) where {N}
+	N == 0 && error("invalid hypot() call without arguments")
+	# speculatively try fast naive approach
+	s = sum(abs2, float.(x))
```

Comment:
`float.(x)` allocates a new tuple, no?  Seems better to do `sum(y -> abs2(float(y)), x)` so that there is a single loop and no temporaries.

Reply:
`Tuple`s are immutable, they are not allocated (on the heap). I tested the performance of these variants,

```julia
julia> fsum(x...) = sum(abs2, float.(x));
julia> hsum(x...) = sum(abs2.(float.(x)));
julia> gsum(x...) = sum(y -> abs2(float(y)), x);

julia> @btime fsum(1.,2.,3.,4.,5.,6.,7.,8.)
  8.116 ns (0 allocations: 0 bytes)
204.0
julia> @btime gsum(1.,2.,3.,4.,5.,6.,7.,8.)
  10.907 ns (0 allocations: 0 bytes)
204.0
julia> @btime hsum(1.,2.,3.,4.,5.,6.,7.,8.)
  2.436 ns (0 allocations: 0 bytes)
204.0
```

Reply:
It seems `hsum` is noticeably faster.

Reply:
Weird, can you file a performance issue for this?  The `gsum` variant should not be so slow.